addon/decompress.h

```diff
@@ -0,0 +1,55 @@
+#include <vector>
+
+#include "compression_worker.h"  // CompressionResult
+#include "zstd.h"
+
+CompressionResult decompress(const std::vector<uint8_t>& compressed) {
+    std::vector<uint8_t> decompressed;
+
+    using DCTX_Deleter = void (*)(ZSTD_DCtx*);
+
+    std::unique_ptr<ZSTD_DCtx, DCTX_Deleter> decompression_context(
+        ZSTD_createDCtx(), [](ZSTD_DCtx* ctx) { ZSTD_freeDCtx(ctx); });
+
+    ZSTD_inBuffer input = {compressed.data(), compressed.size(), 0};
+    std::vector<uint8_t> output_buffer(ZSTD_DStreamOutSize());
+    ZSTD_outBuffer output = {output_buffer.data(), output_buffer.size(), 0};
+
+    //   Source: https://facebook.github.io/zstd/zstd_manual.html#Chapter9
+    //
+    //   Use ZSTD_decompressStream() repetitively to consume your input.
+    //   The function will update both `pos` fields.
+    //   If `input.pos < input.size`, some input has not been consumed.
+    //   It's up to the caller to present again remaining data.
+    //   The function tries to flush all data decoded immediately, respecting output buffer size.
+    //   If `output.pos < output.size`, decoder has flushed everything it could.
+    //   But if `output.pos == output.size`, there might be some data left within internal buffers.,
+    //   In which case, call ZSTD_decompressStream() again to flush whatever remains in the buffer.
+    //   Note : with no additional input provided, amount of data flushed is necessarily <=
+    //   ZSTD_BLOCKSIZE_MAX.
+    //  @return : 0 when a frame is completely decoded and fully flushed,
+    //         or an error code, which can be tested using ZSTD_isError(),
+    //         or any other value > 0, which means there is still some decoding or flushing to do to
+    //         complete current frame :
+    //                                 the return value is a suggested next input size (just a hint
+    //                                 for better latency) that will never request more than the
+    //                                 remaining frame size.
+    auto inputRemains = [](ZSTD_inBuffer& input) { return input.pos < input.size; };
+    auto isOutputBufferFlushed = [](ZSTD_outBuffer& output) { return output.pos < output.size; };
+
+    while (inputRemains(input) || !isOutputBufferFlushed(output)) {
+        size_t const ret = ZSTD_decompressStream(decompression_context.get(), &output, &input);
+        if (ZSTD_isError(ret)) {
+            return std::string(ZSTD_getErrorName(ret));
+        }
+
+        for (size_t i = 0; i < output.pos; ++i) {
+            decompressed.push_back(output_buffer[i]);
+        }
```

Comment:
I considered the latter but decided against it.  It's not as straightforward as maintaining a vector of vectors, because when zstd writes to the output vector, it doesn't update the vector's size, just the bytes (I'm sort of misusing std::vector for the output buffer, perhaps it should instead be a uint8_t[]).  So we'd need to store the vector of decompressed with the number of valid bytes inside it, instead of just the vector.  Alternatively, I could use a uint8[].  But in both cases, we'd need to re-allocate the output buffer per-loop iteration.

I also considered resizing and I am open to this approach, but do you really think this would be more efficient?  resize(), when the new size is greater than the current size, fills the new vector elements with the default value (`0`).  So we'd effectively double iterate the vector for each chunk - once to default construct `output.pos` elements and once more to copy in the decompressed data.

Reply:
> I'm sort of misusing std::vector for the output buffer, perhaps it should instead be a uint8_t[]

I think your usage of std::vector for the output buffer is perfectly fine here 🙂 

> So we'd need to store the vector of decompressed with the number of valid bytes inside it, instead of just the vector.

You could always .resize() after the decompression, right?

> but do you really think this would be more efficient?

... well, yeah: https://godbolt.org/z/oGbzvhco6

Reply:
I switched to resize + copy instead of push_back.